x/datapool/keeper/pool.go

```diff
@@ -553,3 +544,147 @@ func (k Keeper) increaseNumIssuedNFT(ctx sdk.Context, pool *types.Pool) {
 
 	k.SetPool(ctx, pool)
 }
+
+func (k Keeper) RedeemDataPass(ctx sdk.Context, redeemNFT types.MsgRedeemDataPass) (*types.DataPassRedeemReceipt, error) {
+	pool, err := k.GetPool(ctx, redeemNFT.PoolId)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if pool.GetRound() != redeemNFT.Round {
+		return nil, types.ErrRoundNotMatched
+	}
+
+	nftContractAcc, err := sdk.AccAddressFromBech32(pool.NftContractAddr)
+	if err != nil {
+		return nil, sdkerrors.ErrInvalidAddress
+	}
+
+	moduleAddr := types.GetModuleAddress()
+	transferNFTMsg := types.NewTransferNFTMsg(moduleAddr.String(), strconv.FormatUint(redeemNFT.NftId, 10))
+
+	transferNFTMsgBz, err := json.Marshal(transferNFTMsg)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemerAcc, err := sdk.AccAddressFromBech32(redeemNFT.Redeemer)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	redeemTokenIds, err := k.GetRedeemerDataPassByAddr(ctx, pool.PoolId, redeemerAcc)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	if !contains(redeemTokenIds, strconv.FormatUint(redeemNFT.NftId, 10)) {
+		return nil, types.ErrNotOwnedRedeemerNft
+	}
+
+	zeroFund := sdk.NewCoins(types.ZeroFund)
+
+	_, err = k.wasmKeeper.Execute(ctx, nftContractAcc, redeemerAcc, transferNFTMsgBz, zeroFund)
+	if err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrRedeemDataPass, err.Error())
+	}
+
+	nftRedeemReceipt := types.NewDataPassRedeemReceipt(redeemNFT.PoolId, redeemNFT.Round, redeemNFT.NftId, uint64(ctx.BlockHeight()), redeemNFT.Redeemer)
+
```

Comment:
For now, I used `ctx.BlockHeight()`. But in our architecture docs, the `if possible, use block_time` was written. So when all of the other derived branches are merged to this branch, I'll try to change `ctx.BlockTime()`.

Reply:
IIRC, we don't need the block time because we removed the download_period concept (Please correct me if I'm wrong).
So, I think the block height in the DataPassRedeemReceipt is also not mandatory. But, let's keep it for clarity.

Reply:
Right. we don't need block time in redeem receipt anymore.

Reply:
I'll remove `BlockHeight()`. Thanks!